### PR TITLE
[FEATURE] spk setup command: added scaffold HLD and Manifest repos 

### LINKED ISSUE #1086: Add HLD and Manifest repo creation in spk setup
**As a:**
Developer 

**I want:**
HLD and Manifest repos created for me

**So that:**
Can understand how to use SPK with less manual setup

**Does this require updates to documentation?:**
YES

Related to #986

### PR DESCRIPTION
**Pull request by dennisseah**
1. added a request context object to capture all valid input values and status of execution (.e.g is project created, what are the repo created).
2. added `spk setup` as one of the commands.
3. added file operation supports for creating folder, change directory, etc.
4. added git operation supports
5. modified project service to have the right value for `templateTypeId` and `sourceControlType`
6. added scaffold operation supports
7. added code to create setup.log file

main feature added: scaffold HLD and Manifest repos
closes https://github.com/microsoft/bedrock/issues/1086
related to https://github.com/microsoft/bedrock/issues/986

### CODE REVIEW

File: src/commands/setup.md [hunk 1]
```diff
@@ -28,4 +28,27 @@ azdo_project_name=<Azure DevOps Project Name>
 azdo_pat=<Azure DevOps Personal Access Token>
 ```
 
-azdo_project_name is optional and default value is `BedrockRocks`
+`azdo_project_name` is optional and default value is `BedrockRocks`.
+
+The followings shall be created
+
+1. a working directory, `quick-start-env`
```
Comment (edaena) [inline src/commands/setup.md]: nit: capital A in "a working directory"

Comment (dennisseah) [reply src/commands/setup.md]: accept

File: src/commands/setup.ts [hunk 2]
```diff
@@ -4,36 +4,58 @@ import yaml from "js-yaml";
...
 
+export const getErrorMessage = (
+  rc: IRequestContext | undefined,
+  err: Error | IAPIError
+) => {
+  if (rc) {
+    if (err.message && err.message.indexOf("VS402392") !== -1) {
+      return `Project, ${
+        rc!.projectName
+      } might be deleted less than 28 days ago. Choose a different project name.`;
```
Comment (edaena) [inline src/commands/setup.ts]: nit: should it be "might have been"?

Comment (dennisseah) [reply src/commands/setup.ts]: accept

**Review by andrebriggs — APPROVED** [review]

**Comment by samiyaakhtar** [comment]
@dennisseah @andrebriggs  the PR is pretty big and only "related to" the epic. Would be nice if we could break down the epic into smaller tasks and each PR would "close" a task from the epic?

**Comment by yradsmikham** [comment]
@dennisseah @andrebriggs Let's try and keep PRs smaller and manageable.

**Comment by andrebriggs** [comment]
@samiyaakhtar created https://github.com/microsoft/bedrock/issues/1086 and associated with this PR

File: src/commands/setup.md [hunk 3]
```diff
@@ -28,4 +28,27 @@ azdo_project_name=<Azure DevOps Project Name>
...
+
+A `setup.log` file is created after running this command. This file contains
+information about what are created and the execution status (completed or
+incomplete). This file will not be created if input validation failed.
+
+## Notes
+
+1. Personal Access Token. It needs to have the permissions to
+   1. read and write on project
+   1. read and write on Git Repo
```
Comment (mtarng) [inline src/commands/setup.md]: Is this redundant from line 13?

Full required permissions are here: https://github.com/CatalystCode/spk/blob/master/guides/project-service-management-guide.md#generate-personal-access-token

File: src/lib/setup/fsUtil.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import fs from "fs-extra";
+import { removeDir } from "../ioUtil";
+
+/**
+ * Change directory to an absolute path
+ *
+ * @param absPath Absolute Path
+ */
+export const moveToAbsPath = (absPath: string) => {
```
Comment (mtarng) [inline src/lib/setup/fsUtil.ts:9]: Just commenting to say `fsUtil` could probably live in our general lib as could be shared by other commands too.

File: src/lib/setup/gitService.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+): Promise<GitRepository> => {
+  const createOptions = {
+    name: repoName
+  };
+  try {
+    return await gitApi.createRepository(createOptions, projectName);
+  } catch (err) {
+    if (err.statusCode === 401) {
+      throw new Error(
+        `Did not have permissions to vreate git repo. Add git repo write permission to the personal access token`
```
Comment (mtarng) [inline src/lib/setup/gitService.ts]: vreate -> create

**Review by mtarng — APPROVED** [review]
Minor comments. Looks good otherwise.

File: src/commands/setup.md (hunk 3 above)
Comment (dennisseah) [reply src/commands/setup.md]: accept, this is redundant

File: src/lib/setup/fsUtil.ts (hunk 4 above)
Comment (dennisseah) [reply src/lib/setup/fsUtil.ts:9]: we can move it later if needed.

File: src/lib/setup/gitService.ts (hunk 5 above)
Comment (dennisseah) [reply src/lib/setup/gitService.ts]: accept

**Pull request merged**